src/qutip_qip/compiler/gatecompiler.py

```diff
@@ -287,3 +302,87 @@ def _process_idling_tlist(
             # idling until the start time
             idling_tlist.append([start_time])
         return np.concatenate(idling_tlist)
+
+    @classmethod
+    def generate_pulse_shape(cls, shape, num_samples, maximum=1., area=1.):
+        """
+        Return a tuple consisting of a coeff list and a time sequence
+        according to a given pulse shape.
+
+        Parameters
+        ----------
+        shape : str
+            The name ``"rectangular"`` for constant pulse or
+            the name of a Scipy window function.
+            See
+            `the Scipy documentation
+            <https://docs.scipy.org/doc/scipy/reference/signal.windows.html>`_
+            for detail.
+        num_samples : int
+            The number of the samples of the coefficients.
+        maximum : float, optional
+            The maximum of the coefficients.
+            The absolute value will be used if negative.
+        area : float, optional
+            The total area if one integrates coeff as a function of the time.
+            If the area is negative, the pulse is flipped vertically
+            (i.e. the pulse is multiplied by the sign of the area).
+
+        Returns
+        -------
+        coeff, tlist :
+            If the default window ``"shape"="rectangular"`` is used,
+            both are float numbers.
+            If Scipy window functions are used, both are a 1-dimensional numpy
+            array with the same size.
+
+        Notes
+        -----
+        If Scipy window functions are used, it is suggested to set
+        ``Processor.pulse_mode`` to ``"continuous"``.
+        Notice that finite number of sampling points will also make
+        the total integral of the coefficients slightly deviate from ``area``.
+        """
+        coeff, tlist = _normalized_window(shape, num_samples)
+        sign = np.sign(area)
+        coeff *= np.abs(maximum) * sign
+        tlist *= abs(area) / np.abs(maximum)
+        return coeff, tlist
+
+
+"""
+Normalized scipy window functions.
+The scipy implementation only makes sure that it is maximum is 1.
+Here, we save a default t_max so that the integral is always 1.
+"""
```

Comment:
I forgot to mention that this should be in a comment, not a string. A string is only a docstring if it occurs as the first statement in a module, class or function body. Otherwise it is just a string. So this is a bit like write `"a"` on a line by itself in the middle of a program (except more informative, but still a bit strange).